### PR TITLE
add heartbeat package for the websocket connection of the console

### LINKED ISSUE #3293: kubectl terminal session disconnect rapidly
**Describe the Bug**

The container terminal connection uses the WebSocket protocol. WebSockets will be affected by `proxy read timeout` if a reverse proxy is used. We need to increase the timeout or implement a ping message to keep the connection alive. The ping solution has the added benefit of discovering if the connection was closed unexpectedly.

image

**Expected behavior**

Keep the WebSocket session alive.

/assign @wansir @leoendless 
/milestone v3.1.0

### PR DESCRIPTION
**Pull request by yuswift**
Signed-off-by: yuswift <yuswiftli@yunify.com>

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
By adding heartbeat package for websocket connection, we can keep the connection alive so that it will not be killed by the proxy server(e.g. ingress-nginx).
**Which issue(s) this PR fixes**:
Fixes #3293

### CODE REVIEW

File: pkg/models/terminal/terminal.go [hunk 1]
```diff
@@ -200,8 +219,12 @@ func (t *terminaler) HandleSession(shell, namespace, podName, containerName stri
 	var err error
 	validShells := []string{"sh", "bash"}
 
+	stopChan := make(chan struct{})
+
 	session := &TerminalSession{conn: conn, sizeChan: make(chan remotecommand.TerminalSize)}
 
+	go wait.Until(session.sendHeartBeatPackage, time.Duration(3)*time.Second, stopChan)
```
Comment (wansir) [inline pkg/models/terminal/terminal.go:226]: The default interval should be constant, I think the interval should be increased appropriately.

Comment (yuswift) [reply pkg/models/terminal/terminal.go:226]: Should we increase the interval such as 1s, 2s, 4s, 8s?

File: pkg/models/terminal/terminal.go [hunk 2]
```diff
@@ -200,8 +222,12 @@ func (t *terminaler) HandleSession(shell, namespace, podName, containerName stri
 	var err error
 	validShells := []string{"sh", "bash"}
 
+	stopChan := make(chan struct{})
```
Comment (zryfish) [inline pkg/models/terminal/terminal.go:225]: Use stopCha created by apiserver https://github.com/kubesphere/kubesphere/blob/master/cmd/ks-apiserver/app/server.go#L57

File: pkg/models/terminal/terminal.go [hunk 3]
```diff
@@ -36,7 +37,8 @@ import (
 
 const (
 	// Time allowed to write a message to the peer.
-	writeWait = 10 * time.Second
+	writeWait         = 10 * time.Second
+	heartBeatInterval = time.Duration(15) * time.Second
```
Comment (zryfish) [inline pkg/models/terminal/terminal.go:41]:
```suggestion
	heartBeatInterval = 15 * time.Second
```

File: go.mod [hunk 4]
```diff
@@ -96,6 +96,7 @@ require (
 	k8s.io/kube-openapi v0.0.0-20200410145947-61e04a5be9a6
 	k8s.io/kubectl v0.18.6
 	k8s.io/kubernetes v1.14.0
+	k8s.io/metrics v0.18.6
```
Comment (zryfish) [inline go.mod:99]: Why there is a new module imported? I didn't see any code change depended on this module.

Comment (yuswift) [reply go.mod:99]: Me neither. But after I run make `ks-apiserver`, the go.mod & go.sum files are changed. Should we remove the dependency?

File: pkg/models/terminal/terminal.go (hunk 2 above)
Comment (yuswift) [reply pkg/models/terminal/terminal.go:225]: I think the lifecycle of the heartbeat channel should not be controlled by apiserver. If the user closes the terminal page, the server shouldn't send heartbeats anymore, so we close the channel to tell the goroutine to stop.

**Pull request closed without merging**